### ISSUE TITLE
Composite addComponent or addComponents

### ISSUE THREAD

**Issue by las1991** [netty/netty#8411]
@Test
    public void testAddComponent() {
        CompositeByteBuf compositeByteBuf = Unpooled.compositeBuffer();
        compositeByteBuf.writeByte(1);

        ByteBuf byteBuf = Unpooled.buffer(1);
        byteBuf.writeByte(2);
        compositeByteBuf.addComponent(true, byteBuf);
        compositeByteBuf.addComponent(true, byteBuf);
        byteBuf.release();

        compositeByteBuf.writeByte(3);

        System.out.println(ByteBufUtil.prettyHexDump(compositeByteBuf));

        assert compositeByteBuf.getByte(0) == 1;
        assert compositeByteBuf.getByte(2) == 2;
        assert compositeByteBuf.getByte(3) == 2;
        assert compositeByteBuf.getByte(4) == 3;

    }
Test failed !
Expected：

                   +-------------------------------------------------+
                    |  0   1   2   3   4   5   6   7   8   9   a   b   c   d   e   f   |
+----------+-------------------------------------------------+----------------+
|00000000| 01 02 02 03                                                               |....                        |
+----------+-------------------------------------------------+----------------+

Actually :
 
                   +-------------------------------------------------+
                    |  0   1   2   3   4   5   6   7   8   9   a   b   c   d   e   f   |
+----------+-------------------------------------------------+----------------+
|00000000| 01 00 00 03                                                               |....                        |
+----------+-------------------------------------------------+----------------+

I'm not sure I use it correctly.

**Comment by njhill** [comment]
@las1991 the problem is that adding a component to a composite buffer is not the same as writing data to it. When you write the first byte, the buffer contains no components and has no capacity, and therefore implicitly adds some capacity by creating an initial internal `ByteBuf` with a default minimum size of 64 bytes. When you subsequently add the single-byte buffer components, these are appended to the end of the component list, so the capacity is now 66 bytes.

So you could achieve what you want by changing the first write to `compositeBuf.capacity(1).writeByte(1)` or `compositeBuf.writeByte(1).capacity(1)`, which in the first case will add an initial component with capacity one which you can write into, and second will trim the 64-byte component down to 1.

There is some strangeness here w.r.t. passing `increaseWriterIndex` = `true`, since that was only really intended for when the writerIndex == capacity, which is not the case after you have added the first byte (writerIndex = 1, capacity = 64).

@normanmaurer while I assume this kind of usage is generally incorrect, it seems that more "appropriate" behaviour would be for `increaseWriterIndex` to only apply if the insertion _offset_ of the new component is <= the current writerIndex, i.e. would ensure the writer index is still pointing at "the same place" within the data, including if that happens to be the end of it (normal case).

Also what are your thoughts on a method for adding a component at a specific _byte_ index (in addition to the current ones for _component_ index)? In the case that this lies within an existing component, that component would be split with the new one spliced in the middle (i.e. resulting component count might increase by either 1 or 2, but that behaviour could be documented). This would actually be useful for some use cases I have.

Even more general/useful would be such a method which takes both an index and a length. This would perform the appropriate splicing to "replace" the specified region with the buffer to add. The size of the added buffer would not have to match the specified length; the insertion case above would correspond to passing length = 0.

If this sounds desirable I would be happy to submit a PR.

**Comment by normanmaurer** [comment]
@las1991 @njhill I think what should happen is that if you call `addComponent(true...)` and there is already a ByteBuf in there that still has space it should basically "slice" the already existing ByteBuf in the composite before adding the component. 

Is it clear what I have in mind ?

**Comment by njhill** [comment]
@normanmaurer so you are proposing that semantics of `increaseWriterIndex` = `true` should be "insert at writerIndex and move writerIndex to end of inserted buffer", is that correct? This would be a relative-index equivalent of the new method I was proposing.

Some potential problems with this I can see:
- Current semantics of all the non-cIndex versions of `addComponent(...)` is to add the new component at the end. There might be existing code using `addComponent(true, ...)` which doesn't pay attention to the writer index - i.e. absolute usage only or positions the indices explicitly after the composite is constructed. This would break those places since effective final contents of the buffer will change (ignoring the relative read/write indices) ... of course they could fix by changing true to false...
- The meaning of `addComponent(cIndex, buf, true)` would become ambiguous - there are now two possible insertion points (the specified component index and the current writer index)

My suggestion was a smaller "correction" to minimize change to the existing semantics - it does still kind of makes sense to increase the writerIndex by the inserted length if the component it's currently pointing at has been moved by that amount (i.e. is in position < cIndex), but not otherwise.

**Comment by las1991** [comment]
@normanmaurer yes , it will solve the problem , but it would produce much slice component when call write() and addComponent() Interlaced .

**Comment by njhill** [comment]
Looks like this isn't the first time this has been debated :) https://github.com/netty/netty/issues/1196#issuecomment-33870603, but maybe the existing-writer-index confusion wasn't fully considered at the later time that the boolean param was introduced.
